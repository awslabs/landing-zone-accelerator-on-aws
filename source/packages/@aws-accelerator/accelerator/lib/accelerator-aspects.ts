/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { IConstruct } from 'constructs';
import { version } from '../../../../package.json';
import { createLogger } from '@aws-accelerator/utils/lib/logger';
import { getGlobalRegion } from '@aws-accelerator/utils';

const logger = createLogger(['accelerator-aspects']);
/**
 * Property overrides for GovCloud environments
 */
class GovCloudOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('KmsKeyId');
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = node.assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.amazonaws-us-gov.com/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for ISO-E environments
 */
class IsoeOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = node.assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://console.csphome.adc-e.uk/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for ISO-F environments
 */
class IsofOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = node.assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.csphome.hci.ic.gov/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for ISO-B environments
 */
class IsobOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_ec2.CfnFlowLog) {
      node.addPropertyDeletionOverride('LogFormat');
      node.addPropertyDeletionOverride('Tags');
      node.addPropertyDeletionOverride('MaxAggregationInterval');
    }
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('KmsKeyId');
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_cloudtrail.CfnTrail) {
      node.addPropertyDeletionOverride('InsightSelectors');
      node.addPropertyDeletionOverride('IsOrganizationTrail');
    }
    if (node instanceof cdk.aws_ecr.CfnRepository) {
      node.addPropertyDeletionOverride('ImageTagMutability');
    }
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = node.assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.sc2shome.sgov.gov/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for ISO environments
 */
class IsoOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_ec2.CfnFlowLog) {
      node.addPropertyDeletionOverride('LogFormat');
      node.addPropertyDeletionOverride('Tags');
      node.addPropertyDeletionOverride('MaxAggregationInterval');
    }
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('KmsKeyId');
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_cloudtrail.CfnTrail) {
      node.addPropertyDeletionOverride('InsightSelectors');
      node.addPropertyDeletionOverride('IsOrganizationTrail');
    }
    if (node instanceof cdk.aws_ecr.CfnRepository) {
      node.addPropertyDeletionOverride('ImageTagMutability');
    }
    if (node instanceof cdk.aws_iam.CfnRole) {
      const trustPolicyDoc = node.assumeRolePolicyDocument as cdk.aws_iam.SamlConsolePrincipal;
      if (JSON.stringify(trustPolicyDoc).includes('signin.aws.amazon.com')) {
        node.addPropertyOverride(
          'AssumeRolePolicyDocument.Statement.0.Condition.StringEquals.SAML:aud',
          'https://signin.c2shome.ic.gov/saml',
        );
      }
    }
  }
}

/**
 * Property overrides for CN environments
 */
class CnOverrides implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof cdk.aws_logs.CfnLogGroup) {
      node.addPropertyDeletionOverride('Tags');
    }
    if (node instanceof cdk.aws_cloudtrail.CfnTrail) {
      node.addPropertyDeletionOverride('IsOrganizationTrail');
    }
  }
}

/**
 * Default memory override for Lambda resources
 */
class LambdaDefaultMemoryAspect implements cdk.IAspect {
  visit(node: IConstruct): void {
    if (node instanceof cdk.CfnResource) {
      if (node.cfnResourceType === 'AWS::Lambda::Function') {
        const cfnProps = (node as cdk.aws_lambda.CfnFunction)['_cfnProperties'];
        let memorySize = cfnProps['MemorySize']?.toString();

        if (!memorySize) {
          memorySize = (node as cdk.aws_lambda.CfnFunction).memorySize;
        }

        if (!memorySize || memorySize < 512) {
          node.addPropertyOverride('MemorySize', 512);
        }
      }
    }
  }
}
/**
 * Default deletion override for Service linked role resources
 */
class IamServiceLinkedRoleAspect implements cdk.IAspect {
  visit(node: IConstruct): void {
    if (node instanceof cdk.CfnResource) {
      if (node.cfnResourceType === 'AWS::IAM::ServiceLinkedRole') {
        node.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN);
      }
    }
  }
}

/**
 * Solution ID override for Lambda resources
 */
class AwsSolutionAspect implements cdk.IAspect {
  visit(node: IConstruct): void {
    if (node instanceof cdk.CfnResource) {
      if (node.cfnResourceType === 'AWS::Lambda::Function') {
        node.addPropertyOverride('Environment.Variables.SOLUTION_ID', `AwsSolution/SO0199/${version}`);
      }
    }
  }
}

/**
 * Existing Role Overrides
 */
class ExistingRoleOverrides implements cdk.IAspect {
  public visit(construct: IConstruct): void {
    const acceleratorPrefix = process.env['ACCELERATOR_PREFIX'] ?? 'AWSAccelerator';

    if (construct instanceof cdk.CfnResource && construct.cfnResourceType === 'AWS::CloudTrail::Trail') {
      this.replaceCloudTrailCloudWatchLogsRole(construct, acceleratorPrefix);
    } else if (construct instanceof cdk.CfnResource && construct.cfnResourceType === 'AWS::Lambda::Function') {
      this.replaceLambdaFunctionRole(construct, acceleratorPrefix);
    } else if (
      construct instanceof cdk.CfnResource &&
      (construct.cfnResourceType === 'AWS::IAM::Role' ||
        construct.cfnResourceType === 'AWS::IAM::Policy' ||
        construct.cfnResourceType === 'AWS::IAM::InstanceProfile' ||
        construct.cfnResourceType === 'AWS::IAM::ManagedPolicy')
    ) {
      for (const x of construct.obtainResourceDependencies()) {
        construct.removeDependency(x);
      }
      construct.node.scope?.node.tryRemoveChild(construct.node.id);
    }
  }
  private replaceCloudTrailCloudWatchLogsRole(construct: cdk.CfnResource, acceleratorPrefix: string) {
    for (const eachDependency of construct.node.dependencies) {
      // in cloudtrail look for logsRole
      // this is only created when sendToCloudWatchLogs is set to true
      // replace this with pre-existing role
      if (eachDependency.node.path.includes('LogsRole')) {
        construct.addPropertyDeletionOverride('CloudWatchLogsRoleArn');
        construct.addPropertyOverride(
          'CloudWatchLogsRoleArn.Fn::Sub',
          'arn:${AWS::Partition}:iam::${AWS::AccountId}:role/' + acceleratorPrefix + 'CloudTrailCloudWatchRole',
        );
      }
    }
  }
  private replaceLambdaFunctionRole(construct: cdk.CfnResource, acceleratorPrefix: string) {
    for (const x of construct.obtainResourceDependencies()) {
      construct.removeDependency(x);
      const parentConstruct = construct.node.scope;
      parentConstruct?.node.tryRemoveChild(x.node.id);
    }
    construct.addPropertyDeletionOverride('Role');
    construct.addPropertyOverride(
      'Role.Fn::Sub',
      'arn:${AWS::Partition}:iam::${AWS::AccountId}:role/' + acceleratorPrefix + 'LambdaRole',
    );
  }
}

/**
 * Permission boundary aspect
 */
export class PermissionsBoundaryAspect implements cdk.IAspect {
  /**
   * Account this will be applied in
   */
  readonly account: string;
  /**
   * Partition this will be deployed to
   */
  readonly partition: string;
  constructor(account: string, partition: string) {
    this.account = account;
    this.partition = partition;
  }
  public visit(node: IConstruct): void {
    const policyLength = (process.env['ACCELERATOR_PERMISSION_BOUNDARY'] ?? '').trim().length;
    // check if node is type of cloudformation resource
    if (!(node instanceof cdk.CfnResource)) {
      return;
    }
    // check if node is type of IAM role
    if (node.cfnResourceType !== 'AWS::IAM::Role') {
      return;
    }
    // check if deployment is external
    if (process.env['MANAGEMENT_ACCOUNT_ID'] && process.env['MANAGEMENT_ACCOUNT_ROLE_NAME']) {
      return;
    }
    // check if its management account
    if (this.account !== process.env['PIPELINE_ACCOUNT_ID']!) {
      return;
    }
    // policy name is not empty
    if (policyLength === 0) {
      return;
    }

    try {
      // Build permissions boundary ARN from input
      const permissionsBoundaryArn = `arn:${this.partition}:iam::${this.account}:policy/${process.env[
        'ACCELERATOR_PERMISSION_BOUNDARY'
      ]!}`;
      // convert role in to cfn.role this allows for checking properties
      const roleResource = node as cdk.aws_iam.CfnRole;
      if (roleResource && roleResource.permissionsBoundary && roleResource.permissionsBoundary > '') {
        //do nothing, use existing permission boundary
      } else {
        // no permission boundary was found, add permission boundary
        roleResource.addPropertyOverride('PermissionsBoundary', permissionsBoundaryArn);
      }
    } catch (error) {
      const msg = `Error while applying permission boundary to IAM role ${node.node.path}. Error: ${JSON.stringify(
        error,
      )}. Permission boundary will not be applied to stacks.`;
      logger.error(msg);
      throw new Error(msg);
    }
  }
}

/**
 * Add accelerator specific aspects to the application based on partition
 */
export class AcceleratorAspects {
  /**
   * The region for global API endpoints
   * based on AWS partition
   */
  public readonly globalRegion: string;

  constructor(app: cdk.App, partition: string, useExistingRoles: boolean) {
    const globalRegion = getGlobalRegion(partition);
    // Add partition specific overrides
    switch (partition) {
      case 'aws-us-gov':
        cdk.Aspects.of(app).add(new GovCloudOverrides());
        break;
      case 'aws-iso':
        cdk.Aspects.of(app).add(new IsoOverrides());
        break;
      case 'aws-iso-b':
        cdk.Aspects.of(app).add(new IsobOverrides());
        break;
      case 'aws-iso-e':
        cdk.Aspects.of(app).add(new IsoeOverrides());
        break;
      case 'aws-iso-f':
        cdk.Aspects.of(app).add(new IsofOverrides());
        break;
      case 'aws-cn':
        cdk.Aspects.of(app).add(new CnOverrides());
        break;
    }
    // Add default aspects
    cdk.Aspects.of(app).add(new LambdaDefaultMemoryAspect());
    cdk.Aspects.of(app).add(new IamServiceLinkedRoleAspect());
    if (useExistingRoles) {
      cdk.Aspects.of(app).add(new ExistingRoleOverrides());
    } else {
      /**
       * when existing roles are recreated the Lambda, invoke fails on custom resource as environment variable is encrypted and it cannot access it with the error below
       * @example
       * Calling the invoke API action failed with this message: Lambda was unable to decrypt the environment variables because KMS access was denied. Please check the function's KMS key settings. KMS Exception: AccessDeniedExceptionKMS Message: The ciphertext refers to a customer master key that does not exist, does not exist in this region, or you are not allowed to access
       */
      // removing solutions aspect to prevent this error
      cdk.Aspects.of(app).add(new AwsSolutionAspect());
    }
    // Set global region
    this.globalRegion = globalRegion;
  }
}
