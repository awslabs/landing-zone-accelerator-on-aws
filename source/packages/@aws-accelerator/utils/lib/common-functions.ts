import {
  STSClient,
  AssumeRoleCommand,
  AssumeRoleCommandInput,
  AssumeRoleCommandOutput,
  GetCallerIdentityCommand,
} from '@aws-sdk/client-sts';
import { throttlingBackOff } from './throttle';
import { ConfiguredRetryStrategy } from '@aws-sdk/util-retry';
import { createLogger } from './logger';
import { glob } from 'glob';
import { dirname } from 'path';
import * as fs from 'fs';
const logger = createLogger(['utils-common-functions']);

export function chunkArray<Type>(array: Type[], chunkSize: number): Type[][] {
  const chunkedArray: Type[][] = [];
  let index = 0;

  while (index < array.length) {
    chunkedArray.push(array.slice(index, index + chunkSize));
    index += chunkSize;
  }
  return chunkedArray;
}

export async function getCrossAccountCredentials(
  accountId: string,
  region: string,
  partition: string,
  managementAccountAccessRole: string,
  sessionName?: string,
) {
  const stsClient = new STSClient({
    region: region,
    retryStrategy: setRetryStrategy(),
    endpoint: getStsEndpoint(partition, region),
  });
  const stsParams: AssumeRoleCommandInput = {
    RoleArn: `arn:${partition}:iam::${accountId}:role/${managementAccountAccessRole}`,
    RoleSessionName: sessionName ?? 'acceleratorBootstrapCheck',
    DurationSeconds: 3600,
  };
  let assumeRoleCredential: AssumeRoleCommandOutput | undefined = undefined;
  try {
    assumeRoleCredential = await throttlingBackOff(() => stsClient.send(new AssumeRoleCommand(stsParams)));
    if (assumeRoleCredential) {
      return assumeRoleCredential;
    } else {
      throw new Error(`Error assuming role ${managementAccountAccessRole} in account ${accountId} ${sessionName}`);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (e: any) {
    logger.error(JSON.stringify(e));
    throw new Error(e.message);
  }
}

export async function getCurrentAccountId(partition: string, region: string): Promise<string> {
  logger.debug(
    `Sts endpoint for partition ${partition} and region ${region} is : ${getStsEndpoint(partition, region)}`,
  );
  const stsClient = new STSClient({
    region,
    retryStrategy: setRetryStrategy(),
    endpoint: getStsEndpoint(partition, region),
  });
  try {
    const response = await throttlingBackOff(() => stsClient.send(new GetCallerIdentityCommand({})));
    logger.debug(`Current account id is ${response.Account!}`);
    return response.Account!;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    logger.error(`Error to assume role while trying to get current account id ${JSON.stringify(error)}`);
    throw new Error(error.message);
  }
}

export function setRetryStrategy() {
  const numberOfRetries = Number(process.env['ACCELERATOR_SDK_MAX_ATTEMPTS'] ?? 800);
  return new ConfiguredRetryStrategy(numberOfRetries, (attempt: number) => 100 + attempt * 1000);
}

export function getStsEndpoint(partition: string, region: string): string {
  if (partition === 'aws-iso') {
    return `https://sts.${region}.c2s.ic.gov`;
  } else if (partition === 'aws-iso-b') {
    return `https://sts.${region}.sc2s.sgov.gov`;
  } else if (partition === 'aws-cn') {
    return `https://sts.${region}.amazonaws.com.cn`;
  } else if (partition === 'aws-iso-f') {
    return `https://sts.${region}.csp.hci.ic.gov`;
  } else if (partition === 'aws-iso-e') {
    return `https://sts.${region}.cloud.adc-e.uk`;
  }
  // both commercial and govCloud return this pattern
  return `https://sts.${region}.amazonaws.com`;
}

// Converts strings with wildcards (*) to regular expressions to determine if log group name matches exclusion pattern
export function wildcardMatch(text: string, pattern: string): boolean {
  const regexPattern = new RegExp('^' + pattern.replace(/\?/g, '.').replace(/\*/g, '.*') + '$');
  logger.debug(`Converted wildcard pattern ${pattern} to regex pattern ${regexPattern}`);
  const patternMatch = regexPattern.test(text);
  logger.info(`Checking if input string ${text} matches pattern ${pattern} provided. Result: ${patternMatch}`);
  return patternMatch;
}

/**
 * Returns STS credentials for a given role ARN
 * @param stsClient STSClient
 * @param roleArn string
 * @returns `Promise<{ accessKeyId: string; secretAccessKey: string; sessionToken: string }>`
 */
export async function getStsCredentials(
  stsClient: STSClient,
  roleArn: string,
): Promise<{ accessKeyId: string; secretAccessKey: string; sessionToken: string }> {
  console.log(`Assuming role ${roleArn}...`);
  try {
    const response = await throttlingBackOff(() =>
      stsClient.send(new AssumeRoleCommand({ RoleArn: roleArn, RoleSessionName: 'AcceleratorAssumeRole' })),
    );
    //
    // Validate response
    if (!response.Credentials?.AccessKeyId) {
      throw new Error(`Access key ID not returned from AssumeRole command`);
    }
    if (!response.Credentials.SecretAccessKey) {
      throw new Error(`Secret access key not returned from AssumeRole command`);
    }
    if (!response.Credentials.SessionToken) {
      throw new Error(`Session token not returned from AssumeRole command`);
    }

    return {
      accessKeyId: response.Credentials.AccessKeyId,
      secretAccessKey: response.Credentials.SecretAccessKey,
      sessionToken: response.Credentials.SessionToken,
    };
  } catch (e) {
    throw new Error(`Could not assume role: ${e}`);
  }
}

/**
 * Function getAllFilesInPattern - returns all file names in a given directory that match a pattern
 * its recursive so top level files will be returned as well as files in sub directories
 * Only file name will be returned. For example, if file is testName.extension then return will be testName
 */
export async function getAllFilesInPattern(dir: string, pattern: string, fullPath?: boolean): Promise<string[]> {
  const files = await glob(`${dir}/**/*${pattern}`, {
    ignore: ['**/node_modules/**'],
  });
  // logger.debug(`Found ${JSON.stringify(files)} files matching pattern ${pattern}`);
  const parsedFiles = files.map(file => {
    return file.replace(dirname(file), '').replace('/', '').replace(pattern, '');
  });
  // logger.debug(`Parsed files ${JSON.stringify(parsedFiles)}`);
  if (fullPath) {
    return files;
  }
  return parsedFiles;
}

export async function checkDiffFiles(dir: string, templatePattern: string, diffPattern: string) {
  const templateFiles = await getAllFilesInPattern(dir, templatePattern);
  const diffFiles = await getAllFilesInPattern(dir, diffPattern);
  if (templateFiles.length !== diffFiles.length) {
    throw new Error(
      `Number of template files ${templateFiles.length} does not match number of diff files ${diffFiles.length} in directory ${dir}`,
    );
  }
}

/**
 * Sets the global region for API calls based on the given partition
 * @param partition
 * @returns region
 */
export function getGlobalRegion(partition: string): string {
  switch (partition) {
    case 'aws-us-gov':
      return 'us-gov-west-1';
    case 'aws-iso':
      return 'us-iso-east-1';
    case 'aws-iso-b':
      return 'us-isob-east-1';
    case 'aws-iso-e':
      return 'eu-isoe-west-1';
    case 'aws-iso-f':
      return 'us-isof-south-1';
    case 'aws-cn':
      return 'cn-northwest-1';
    default:
      return 'us-east-1';
  }
}

export async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.promises.stat(filePath);
    return true;
  } catch (err) {
    return false;
  }
}

export async function directoryExists(directory: string): Promise<boolean> {
  try {
    const stats = await fs.promises.stat(directory);
    return stats.isDirectory();
  } catch (err) {
    return false;
  }
}
