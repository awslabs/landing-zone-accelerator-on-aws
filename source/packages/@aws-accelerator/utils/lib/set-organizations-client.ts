/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { getGlobalRegion, setRetryStrategy } from './common-functions';
import { OrganizationsClient } from '@aws-sdk/client-organizations';

/**
 * Sets an SDKv3 Organizations client based on partition
 * @param partition string
 * @param solutionId string | undefined
 * @returns OrganizationsClient
 */
export function setOrganizationsClient(partition: string, solutionId?: string): OrganizationsClient {
  return new OrganizationsClient({
    region: getGlobalRegion(partition),
    customUserAgent: solutionId,
    retryStrategy: setRetryStrategy(),
  });
}
