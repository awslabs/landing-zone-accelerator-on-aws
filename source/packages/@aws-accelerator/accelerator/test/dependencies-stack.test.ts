/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { AcceleratorStage } from '../lib/accelerator-stage';
import { describe, test } from '@jest/globals';
import { snapShotTest } from './snapshot-test';
import { Template } from 'aws-cdk-lib/assertions';
import { Create, memoize } from './accelerator-test-helpers';

describe('DependenciesStack', () => {
  snapShotTest(
    'Construct(DependenciesStack): ',
    Create.stackProvider('Management-us-east-1', AcceleratorStage.DEPENDENCIES),
  );
});

const getMultiOuStack = memoize(
  Create.stackProvider('Management-us-east-1', [
    AcceleratorStage.DEPENDENCIES,
    'aws',
    'us-east-1',
    'all-enabled-ou-targets',
  ]),
);

describe('default event bus policy', () => {
  test('default event bus policy is created', () => {
    const multiOuStack = getMultiOuStack()!;
    const template = Template.fromStack(multiOuStack);

    template.hasResourceProperties('AWS::Events::EventBusPolicy', { EventBusName: 'default' });
    template.resourceCountIs('AWS::Events::EventBusPolicy', 4);
  });
});
