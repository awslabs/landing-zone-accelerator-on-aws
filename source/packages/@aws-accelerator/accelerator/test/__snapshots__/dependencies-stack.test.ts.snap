// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependenciesStack Construct(DependenciesStack):  Snapshot Test 1`] = `
{
  "Resources": {
    "EventBusPolicy0": {
      "Properties": {
        "EventBusName": "default",
        "Statement": {
          "Action": [
            "events:DisableRule",
            "events:DeleteRule",
          ],
          "Condition": {
            "ForAllValues:StringNotLike": {
              "aws:PrincipalArn": [
                "arn:aws:iam::*:role/AWSControlTowerExecution",
                "arn:aws:iam::*:role/AWSAccelerator*",
                "arn:aws:iam::*:role/cdk-accel-*",
                "arn:aws:iam::*:role/aws-service-role/*.amazonaws.com/AWSServiceRole*",
              ],
            },
          },
          "Effect": "Deny",
          "Principal": "*",
          "Resource": [
            "arn:aws:events:us-east-1:111111111111:rule/*",
          ],
        },
        "StatementId": "DenyDisableRulePolicy",
      },
      "Type": "AWS::Events::EventBusPolicy",
    },
    "EventBusPolicy1": {
      "Properties": {
        "EventBusName": "default",
        "Statement": {
          "Action": [
            "events:PutEvents",
            "events:PutTargets",
            "events:PutRule",
          ],
          "Condition": {
            "Bool": {
              "aws:PrincipalIsAWSService": "true",
            },
            "StringEquals": {
              "aws:PrincipalOrgID": "o-asdf123456",
            },
          },
          "Effect": "Allow",
          "Principal": "*",
          "Resource": [
            "arn:aws:events:us-east-1:111111111111:event-bus/default",
          ],
        },
        "StatementId": "AwsServivePutEventsPolicy",
      },
      "Type": "AWS::Events::EventBusPolicy",
    },
    "EventBusPolicy2": {
      "Properties": {
        "EventBusName": "default",
        "Statement": {
          "Action": [
            "events:DisableRule",
            "events:DeleteRule",
          ],
          "Condition": {
            "StringEquals": {
              "aws:PrincipalOrgID": "o-asdf123456",
            },
          },
          "Effect": "Allow",
          "Principal": "*",
          "Resource": [
            "arn:aws:events:us-east-1:111111111111:rule/*",
          ],
        },
        "StatementId": "AllowOrgPolicy",
      },
      "Type": "AWS::Events::EventBusPolicy",
    },
    "EventBusPolicy3": {
      "Properties": {
        "EventBusName": "default",
        "Statement": {
          "Action": [
            "events:PutEvents",
            "events:PutTargets",
            "events:PutRule",
          ],
          "Condition": {
            "ForAllValues:StringNotLike": {
              "aws:PrincipalArn": [
                "arn:aws:iam::*:role/AWSControlTowerExecution",
                "arn:aws:iam::*:role/AWSAccelerator*",
                "arn:aws:iam::*:role/cdk-accel-*",
                "arn:aws:iam::*:role/aws-service-role/*.amazonaws.com/AWSServiceRole*",
              ],
            },
          },
          "Effect": "Deny",
          "Principal": "*",
          "Resource": [
            "arn:aws:events:us-east-1:111111111111:event-bus/default",
          ],
        },
        "StatementId": "AllowPutEventsPolicy",
      },
      "Type": "AWS::Events::EventBusPolicy",
    },
    "LzaManagedPolicy01FDAB5303": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Policies definition are derived from accelerator iam-config boundary-policy file",
            },
          ],
        },
      },
      "Properties": {
        "Description": "",
        "ManagedPolicyName": "lzaManagedPolicy01",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:CreateBucket",
                "s3:PutBucketPolicy",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::rcc-bucket-trustedservice-do-not-delete-*",
            },
            {
              "Action": "s3:ListAllMyBuckets",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::*",
            },
            {
              "Action": [
                "config:DescribeConfigurationRecorders",
                "config:DescribeConfigurationRecorderStatus",
                "config:PutConfigurationRecorder",
                "config:StartConfigurationRecorder",
                "config:DescribeDeliveryChannels",
                "config:DescribeDeliveryChannelStatus",
                "config:PutDeliveryChannel",
                "config:DescribeAggregationAuthorizations",
                "config:PutAggregationAuthorization",
                "config:ListDiscoveredResources",
                "config:BatchGetResourceConfig",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "iam:GetRole",
                "iam:PassRole",
                "iam:CreateServiceLinkedRole",
                "iam:DeleteServiceLinkedRole",
                "iam:GetServiceLinkedRoleDeletionStatus",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:iam::*:role/aws-service-role/config.amazonaws.com/*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "LzaManagedPolicy027214A961": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Policies definition are derived from accelerator iam-config boundary-policy file",
            },
          ],
        },
      },
      "Properties": {
        "Description": "",
        "ManagedPolicyName": "lzaManagedPolicy02",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:CreateBucket",
                "s3:PutBucketPolicy",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::rcc-bucket-trustedservice-do-not-delete-*",
            },
            {
              "Action": "s3:ListAllMyBuckets",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::*",
            },
            {
              "Action": [
                "config:DescribeConfigurationRecorders",
                "config:DescribeConfigurationRecorderStatus",
                "config:PutConfigurationRecorder",
                "config:StartConfigurationRecorder",
                "config:DescribeDeliveryChannels",
                "config:DescribeDeliveryChannelStatus",
                "config:PutDeliveryChannel",
                "config:DescribeAggregationAuthorizations",
                "config:PutAggregationAuthorization",
                "config:ListDiscoveredResources",
                "config:BatchGetResourceConfig",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "iam:GetRole",
                "iam:PassRole",
                "iam:CreateServiceLinkedRole",
                "iam:DeleteServiceLinkedRole",
                "iam:GetServiceLinkedRoleDeletionStatus",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:iam::*:role/aws-service-role/config.amazonaws.com/*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "PutSsmParameterRoleEF99BE78": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM5",
              "reason": "This role is required to give permissions to put/delete SSM parameters across accounts and regions",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "ArnLike": {
                  "aws:PrincipalArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::*:role/AWSAccelerator*",
                      ],
                    ],
                  },
                },
                "StringEquals": {
                  "aws:PrincipalOrgID": "o-asdf123456",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ssm:PutParameter",
                    "ssm:DeleteParameter",
                    "ssm:GetParameter",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:ssm:*:*:parameter/accelerator*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "default",
          },
        ],
        "RoleName": "AWSAccelerator-CrossAccountSsmParameterShare",
      },
      "Type": "AWS::IAM::Role",
    },
    "SsmParamAcceleratorVersionFF83282D": {
      "Properties": {
        "Name": "/accelerator/AWSAccelerator-DependenciesStack-111111111111-us-east-1/version",
        "Type": "String",
        "Value": "1.11.1",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SsmParamStackId521A78D3": {
      "Properties": {
        "Name": "/accelerator/AWSAccelerator-DependenciesStack-111111111111-us-east-1/stack-id",
        "Type": "String",
        "Value": {
          "Ref": "AWS::StackId",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
